In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

^C


### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
0 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
9

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Overwriting sum.py


In [ ]:
!python sum.py

Traceback (most recent call last):
  File "/content/sum.py", line 7, in <module>
    a, b = map(int, input().split())
KeyboardInterrupt
^C


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [22]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [23]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

^C


In [19]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product.cpp


### Test 1: Small numbers

In [16]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Writing input0.txt


In [17]:
%%writefile output0.txt
0

Writing output0.txt


In [25]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [27]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Writing input1.txt


In [28]:
!./max_pairwise_product_full < input1.txt

90


In [ ]:
%%writefile output1.txt
90

Overwriting output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [83]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Writing input2.txt


In [ ]:
%%writefile output2.txt
90

Overwriting output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [95]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [ ]:
200000*200000

40000000000

In [ ]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

1345294336


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [5]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


### Changing the variables to long long

In [37]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long_full.cpp


In [41]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [43]:
!./max_pairwise_product_long_full < input1.txt

90


### Runing multuple test with a bash shell

In [32]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [33]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

./run_all_test.sh: line 5: input{1}.txt: No such file or directory
grep: output{1}.txt: No such file or directory


### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [1]:
%%writefile input4.txt
5
300000 41 300000 9 33

Writing input4.txt


In [14]:
300000 * 300000

90000000000

In [44]:
!./max_pairwise_product_long_full < input4.txt

90000000000


In [2]:
%%writefile input5.txt
11
3456 7456 252346 463745673 1 3 0 0 2 5 234

Writing input5.txt


In [13]:
463745673 * 252346

117024365598858

In [45]:
!./max_pairwise_product_long_full < input5.txt

117024365598858


In [3]:
%%writefile input6.txt
4
40000 41 4567 9000

Writing input6.txt


In [12]:
40000 * 9000

360000000

In [46]:
!./max_pairwise_product_long_full < input6.txt

360000000


In [4]:
%%writefile input7.txt
3
12345 2345 654342

Writing input7.txt


In [11]:
12345 * 654342

8077851990

In [47]:
!./max_pairwise_product_long_full < input7.txt

8077851990


In [5]:
%%writefile input8.txt
9
1 2 3 4 5 6 7 8 9

Writing input8.txt


In [10]:
9 * 8

72

In [48]:
!./max_pairwise_product_long_full < input8.txt

72


In [6]:
%%writefile input9.txt
6
56 7865 2345 52 1234 1

Writing input9.txt


In [9]:
7865 * 2345

18443425

In [49]:
!./max_pairwise_product_long_full < input9.txt

18443425


In [50]:
%%writefile input10.txt
102
345678912 987654321 10345 15678 20891 25467 3021 34987 40123 45789 50321 54890 60123 65389 70894 75231 80456 85901 90324 9586 100234 105678 110394 115203 120947 125678 130492 135678 140903 145632 150987 155321 160834 165492 170987 175234 180903 185672 190345 195678 200349 205874 210983 215694 220987 225364 230908 235764 240395 245678 250394 255987 260934 265378 270908 275643 280987 285430 290867 295231 300942 305678 310945 315678 320394 325890 330987 335401 340982 345678 350930 355789 360987 365231 370984 375943 380934 385671 390123 395678 400349 405908 410234 415678 420397 425890 430987 435120 440932 445678 450301 455908 460934 465874 470398 475908 480934 485231 490834 495678 500937 505890

Overwriting input10.txt


In [51]:
345678912 * 98765432

34141127076969984

In [52]:
!./max_pairwise_product_long_full < input10.txt

341411271115378752


### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [53]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [54]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [55]:
199999 * 200000

39999800000

In [56]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [57]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [58]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [59]:
%%writefile output12.txt
39999800000

Writing output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [61]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

39999800000

real	4m56.803s
user	4m54.794s
sys	0m0.055s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	4m58.630s
user	4m55.509s
sys	0m0.084s


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [62]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


### Count executed intructions in code

In [63]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

571


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [ ]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

44
44
601



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [ ]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

572


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

200003177092


In [ ]:
# Estimated time by instruction
# use the real reported  !time ./max_pairwise_product_long_full< input11.txt in seconds intead of the 295

295/200003177092


1.4749765693187072e-09

In [ ]:
# Estimated number of instructions by second

200003177092/295

677976871.4983051

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [64]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
10999 * 11000

120989000

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.859s
user	0m0.854s
sys	0m0.001s


## Python

In [65]:
%%writefile max_pairwise_product_py.py
def max_pairwise_product(l):
    max = 0
    aux = 0
    n = len(l)
    for i in range(n):
        for j in range(n):
            if i != j:
                aux = l[i] * l[j]
                if aux > max:
                    max = aux
    return max

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product(l))

Writing max_pairwise_product_py.py


In [66]:
%%writefile WriteTestPython.py
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("10000\n")
        for i in range(1, 10001):
            f.write(f"{i} ")

Writing WriteTestPython.py


In [67]:
%%time
!python max_pairwise_product_py.py < input14.txt

/bin/bash: line 1: input14.txt: No such file or directory
CPU times: user 5.45 ms, sys: 542 µs, total: 5.99 ms
Wall time: 106 ms


# For the follwing solutions rquired  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [68]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Overwriting max_pairwise_product.cpp


In [69]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_long.cpp


In [72]:
%%writefile max_pairwise_product2_py.py
#Python3
def max_pairwise_product(l):
    max = 0
    aux = 0
    n = len(l)
    for i in range(n):
        for j in range(i + 1, n):
            auxt = l[i] * l[j]
            if max < aux:
                max = aux

    return max

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product(l))

Writing max_pairwise_product2_py.py


In [70]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input11.txt

39999800000

real	2m19.223s
user	2m18.423s
sys	0m0.033s


In [73]:
%%time
!python max_pairwise_product2_py.py < input11.txt

Traceback (most recent call last):
  File "/content/max_pairwise_product2_py.py", line 17, in <module>
    print(max_pairwise_product(l))
  File "/content/max_pairwise_product2_py.py", line 9, in max_pairwise_product
    if max < aux:
KeyboardInterrupt
^C
CPU times: user 3.77 s, sys: 525 ms, total: 4.29 s
Wall time: 14min 53s


El código se pausó debido a que estaba tomando mucho tiempo.

### Estimate and verify the maximun n that can be solved in 1 second



In [77]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  ofstream MyFile("input13.txt");

 MyFile << 16000 <<"\n";  // n
 for (int i=1; i <16001; i++){
     MyFile << i <<" ";
 }
  MyFile.close();
}

Overwriting WriteTest.cpp


In [75]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [78]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input13.txt

^C


# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

In [79]:
%%writefile max_pairwise_product_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main(int argc, char* argv[]){
    int n;
    cin >> n;
    vector<int> v(n);
    for(int i = 0; i < n; i++){
        std::cin >> v[i];
    }

    sort(v.begin(), v.end());
    cout << (long long) v[n - 1] * v[n - 2] << '\n';
    return 0;
}

Writing max_pairwise_product_sort.cpp


In [80]:
%%writefile max_pairwise_product_sort.py
if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    l.sort()
    print(l[-1] * l[-2])

Writing max_pairwise_product_sort.py


Test with inputs 1 and 2



In [84]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort

!./max_pairwise_product_sort < input1.txt
!./max_pairwise_product_sort < input2.txt

90
90


In [85]:
!python max_pairwise_product_sort.py < input1.txt
!python max_pairwise_product_sort.py < input2.txt

90
90


Test and time it with input 3

In [96]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!time ./max_pairwise_product_long < input3.txt

40000000000

real	0m0.003s
user	0m0.003s
sys	0m0.000s


In [97]:
%%time
!python max_pairwise_product_sort.py < input3.txt

40000000000
CPU times: user 4.86 ms, sys: 876 µs, total: 5.74 ms
Wall time: 105 ms


20 stress tests

In [89]:
%%writefile stress_test_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>

int main() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> distrib(2, 1001);

    for (int i = 0; i < 20; ++i) {
        int n = distrib(gen) % 49 + 2;
        std::vector<int> v(n);
        for (int j = 0; j < n; ++j) {
            v[j] = distrib(gen);
        }

        std::sort(v.begin(), v.end(), std::greater<int>());
        long long expected_product = (long long)v[0] * v[1];

        std::cout << "n: " << n << " Expected Product: " << expected_product << '\n';
    }
    return 0;
}

Writing stress_test_cpp.cpp


In [90]:
!g++ stress_test_cpp.cpp -o stress_test_cpp
!./stress_test_cpp

n: 46 Expected Product: 969168
n: 21 Expected Product: 919485
n: 49 Expected Product: 943702
n: 20 Expected Product: 978120
n: 36 Expected Product: 828854
n: 9 Expected Product: 779381
n: 9 Expected Product: 729366
n: 16 Expected Product: 805779
n: 13 Expected Product: 850751
n: 45 Expected Product: 948627
n: 23 Expected Product: 918237
n: 37 Expected Product: 935080
n: 8 Expected Product: 791931
n: 13 Expected Product: 687536
n: 49 Expected Product: 954528
n: 45 Expected Product: 952567
n: 20 Expected Product: 874962
n: 49 Expected Product: 998998
n: 24 Expected Product: 943806
n: 6 Expected Product: 567006


In [91]:
%%writefile stress_tests_sort_py.py
import random

def stress_test():
    for _ in range(20):
        n = random.randint(2, 50)
        arr = [random.randint(2, 1001) for _ in range(n)]

        arr.sort(reverse=True)
        expected_product = arr[0] * arr[1]

        print(f"n: {n}, Expected Product: {expected_product}")

if __name__ == "__main__":
    stress_test()

Writing stress_tests_sort_py.py


In [92]:
!python stress_tests_sort_py.py

n: 12, Expected Product: 808500
n: 2, Expected Product: 9646
n: 35, Expected Product: 849859
n: 49, Expected Product: 946665
n: 13, Expected Product: 969240
n: 46, Expected Product: 963160
n: 28, Expected Product: 918380
n: 31, Expected Product: 962357
n: 39, Expected Product: 982072
n: 19, Expected Product: 849142
n: 15, Expected Product: 817578
n: 44, Expected Product: 948675
n: 29, Expected Product: 979038
n: 35, Expected Product: 898604
n: 35, Expected Product: 852489
n: 6, Expected Product: 968247
n: 16, Expected Product: 951000
n: 11, Expected Product: 488631
n: 41, Expected Product: 960231
n: 8, Expected Product: 374166


Count executed instructions in code and verify with inputs 1,2 and 3


In [93]:
%%writefile max_pairwise_product_sort_ins.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

int main() {
    uint64_t ins = 0;
    int n;
    ins++;
    std::cin >> n;
    ins++;
    std::vector<uint64_t> v(n);
    ins += n;

    for (int i = 0; i < n; ++i) {
        ins++;
        std::cin >> v[i];
        ins++;
    }

    sort(v.begin(), v.end(), greater<uint64_t>());
    ins += n * log2(n);

    uint64_t max_product = (long long)v[0] * v[1];
    ins += 3;

    std::cout << max_product << '\n';
    ins++;

    std::cout << "Executed instructions: " << ins << '\n';
    ins++;

    return 0;
}

Writing max_pairwise_product_sort_ins.cpp


In [98]:
!g++ max_pairwise_product_sort_ins.cpp -o max_pairwise_product_sort_ins
!./max_pairwise_product_sort_ins < input1.txt
!./max_pairwise_product_sort_ins < input2.txt
!./max_pairwise_product_sort_ins < input3.txt

90
Executed instructions: 69
90
Executed instructions: 69
40000000000
Executed instructions: 39


In [103]:
%%writefile max_pairwise_product_sort_ins.py
import math

def max_pairwise_product_sort_ins(arr):
    ins = 0
    n = len(arr)
    ins += 1

    arr.sort(reverse=True)
    ins += n * math.log2(n)

    max_product = arr[0] * arr[1]
    ins += 3

    print(max_product)
    ins += 1

    print("Estimated executed instructions:", int(ins))
    return ins

if __name__ == "__main__":
    ins = 0
    n = int(input())
    ins += 1
    l = list(map(int, input().split()))
    ins += 5 * n
    ins = max_pairwise_product_sort_ins(l)

Overwriting max_pairwise_product_sort_ins.py


In [104]:
!python max_pairwise_product_sort_ins.py < input1.txt
!python max_pairwise_product_sort_ins.py < input2.txt
!python max_pairwise_product_sort_ins.py < input3.txt

90
Estimated executed instructions: 38
90
Estimated executed instructions: 38
40000000000
Estimated executed instructions: 20


In [108]:
%%time
!python max_pairwise_product_sort_ins.py < input10.txt

341411271115378752
Estimated executed instructions: 685
CPU times: user 2.93 ms, sys: 1.98 ms, total: 4.91 ms
Wall time: 105 ms


Calculate estimated time by instruction and number of instructions by second

In [110]:
instructions_per_second  = 685 / 0.00491
time_per_instruction = 0.00491 / 685
print("Estimated time by instruction:", time_per_instruction)
print("Estimated number of instructions by second:", instructions_per_second)

Estimated time by instruction: 7.167883211678832e-06
Estimated number of instructions by second: 139511.2016293279


### Estimate and verify the maximun n that can be solved in 1 second

In [111]:
%%writefile WriteTest.cpp
#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  ofstream MyFile("input13.txt");

  int estimated_n = 1000000;

  MyFile << estimated_n << "\n";

  for (int i = 1; i <= estimated_n; ++i) {
      MyFile << i << " ";
  }

  MyFile.close();

  return 0;
}

Overwriting WriteTest.cpp


In [112]:
!g++ WriteTest.cpp -o WriteTest

In [113]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!time ./max_pairwise_product_sort < input13.txt

26895836000

real	0m0.168s
user	0m0.154s
sys	0m0.002s


In [114]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    estimated_n = 1000000

    with open("input16.txt", "w") as f:
        f.write(f"{estimated_n}\n")
        for i in range(1, estimated_n + 1):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [115]:
!python WriteTestPython.py

In [116]:
%%time
!python max_pairwise_product_sort.py < input5.txt

117024365598858
CPU times: user 5.78 ms, sys: 855 µs, total: 6.63 ms
Wall time: 107 ms


## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [117]:
%%writefile max_pairwise_product_ef.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long max_pairwise_product_f(const std::vector<int>& v) {
    int n = v.size();
    if (n < 2) {
        return 0;
    }

    long long max1 = 0;
    long long max2 = 0;
    int max1_index = -1;

    for (int i = 0; i < n; ++i) {
        if (v[i] > max1) {
            max1 = v[i];
            max1_index = i;
        }
    }

    for (int i = 0; i < n; ++i) {
        if (i != max1_index && v[i] > max2) {
            max2 = v[i];
        }
    }

    return max1 * max2;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> v(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> v[i];
    }
    std::cout << max_pairwise_product_f(v) << '\n';
    return 0;
}

Writing max_pairwise_product_ef.cpp


In [118]:
%%writefile max_pairwise_product_ef.py
def max_pairwise_product_f(numbers):
    n = len(numbers)
    if n < 2:
        return 0

    max1 = 0
    max2 = 0
    max1_index = -1

    for i in range(n):
        if numbers[i] > max1:
            max1 = numbers[i]
            max1_index = i

    for i in range(n):
        if i != max1_index and numbers[i] > max2:
            max

Writing max_pairwise_product_ef.py


Test with inputs 1 and 2



In [119]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!./max_pairwise_product_ef < input1.txt
!./max_pairwise_product_ef < input2.txt

90
90


In [122]:
!python max_pairwise_product_ef.py < input1.txt
!python max_pairwise_product_ef.py < input2.txt

Test and time it with input 3

In [125]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!time ./max_pairwise_product_ef < input3.txt

40000000000

real	0m0.003s
user	0m0.001s
sys	0m0.002s


In [126]:
%%time
!python max_pairwise_product_ef.py < input3.txt

CPU times: user 5.2 ms, sys: 250 µs, total: 5.45 ms
Wall time: 106 ms


20 stress tests

In [127]:
%%writefile stress_tests_ef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>

using namespace std;

long long ins = 0;

long long max_pairwise_product_f(const vector<int>& v) {
    int n = v.size();
    ins++;

    if (n < 2) {
        return 0;
    }

    long long max1 = 0;
    ins++;
    long long max2 = 0;
    ins++;
    int max1_index = -1;


    for (int i = 0; i < n; ++i) {
        ins += 2;
        if (v[i] > max1) {
            max1 = v[i];
            max1_index = i;
            ins += 2;
        }
    }

    for (int i = 0; i < n; ++i) {
        ins += 2;
        if (i != max1_index && v[i] > max2) {
            max2 = v[i];
            ins++;
        }
    }

    ins++;
    return max1 * max2;
}

int main() {
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> distrib(2, 1001);

    for (int i = 0; i < 20; ++i) {
        int n = distrib(gen) % 83 + 2;
        vector<int> v(n);
        for (int j = 0; j < n; ++j) {
            v[j] = distrib(gen);
        }
        ins = 0;
        long long product = max_pairwise_product_f(v);
        cout << "n: " << n << ", Product: " << product << ", Estimated Instructions: " << ins << endl;
    }

    return 0;
}

Writing stress_tests_ef_cpp.cpp


In [128]:
!g++ stress_tests_ef_cpp.cpp -o stress_tests_ef_cpp
!./stress_tests_ef_cpp

n: 46, Product: 888300, Estimated Instructions: 197
n: 64, Product: 935550, Estimated Instructions: 272
n: 69, Product: 960364, Estimated Instructions: 287
n: 33, Product: 840600, Estimated Instructions: 143
n: 27, Product: 949468, Estimated Instructions: 123
n: 57, Product: 948555, Estimated Instructions: 245
n: 7, Product: 834240, Estimated Instructions: 38
n: 43, Product: 913815, Estimated Instructions: 183
n: 27, Product: 974120, Estimated Instructions: 120
n: 21, Product: 865737, Estimated Instructions: 102
n: 56, Product: 960204, Estimated Instructions: 248
n: 22, Product: 811923, Estimated Instructions: 100
n: 38, Product: 916864, Estimated Instructions: 176
n: 27, Product: 906326, Estimated Instructions: 121
n: 75, Product: 968207, Estimated Instructions: 317
n: 75, Product: 935085, Estimated Instructions: 312
n: 34, Product: 963336, Estimated Instructions: 159
n: 69, Product: 982000, Estimated Instructions: 300
n: 36, Product: 949308, Estimated Instructions: 157
n: 18, Product

In [129]:
%%writefile stress_tests_ef_py.py
#Python3
import random

def max_pairwise_product_f(numbers):
    ins = 0
    n = len(numbers)
    ins += 1

    if n < 2:
        return 0, ins

    max1 = 0
    ins += 1
    max2 = 0
    ins += 1
    max1_index = -1

    for i in range(n):
        ins += 2
        if numbers[i] > max1:
            max1 = numbers[i]
            max1_index = i
            ins += 2

    for i in range(n):
        ins += 2
        if i != max1_index and numbers[i] > max2:
            max2 = numbers[i]
            ins += 1
    ins += 1
    return max1 * max2, ins

if __name__ == "__main__":
    for _ in range(20):
        n = random.randint(2, 99)
        numbers = [random.randint(17, 24531) for _ in range(n)]
        product, ins_count = max_pairwise_product_f(numbers)
        print(f"n: {n}, Product: {product}, Estimated Instructions: {ins_count}")

Writing stress_tests_ef_py.py


In [130]:
!python stress_tests_ef_py.py

n: 57, Product: 576497250, Estimated Instructions: 244
n: 84, Product: 571232844, Estimated Instructions: 353
n: 59, Product: 581866246, Estimated Instructions: 254
n: 5, Product: 369273564, Estimated Instructions: 27
n: 51, Product: 584768448, Estimated Instructions: 219
n: 39, Product: 461044935, Estimated Instructions: 168
n: 65, Product: 578691264, Estimated Instructions: 279
n: 53, Product: 561332584, Estimated Instructions: 222
n: 7, Product: 243103428, Estimated Instructions: 40
n: 64, Product: 578790575, Estimated Instructions: 268
n: 40, Product: 528654425, Estimated Instructions: 191
n: 51, Product: 539887360, Estimated Instructions: 228
n: 16, Product: 457681509, Estimated Instructions: 78
n: 74, Product: 579306765, Estimated Instructions: 314
n: 61, Product: 576331664, Estimated Instructions: 262
n: 23, Product: 545383586, Estimated Instructions: 104
n: 72, Product: 579045226, Estimated Instructions: 307
n: 62, Product: 592044624, Estimated Instructions: 273
n: 34, Product:

Count executed instructions in code and verify with inputs 1,2 and 3


In [131]:
%%writefile max_pairwise_product_ef_ins.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

long long ins = 0;

long long max_pairwise_product_f(const vector<int>& v) {
    int n = v.size();
    ins++;

    if (n < 2) {
        return 0;
    }

    long long max1 = 0;
    ins++;
    long long max2 = 0;
    ins++;
    int max1_index = -1;

    for (int i = 0; i < n; ++i) {
        ins += 2;
        if (v[i] > max1) {
            max1 = v[i];
            max1_index = i;
            ins += 2;
        }
    }
    for (int i = 0; i < n; ++i) {
        ins += 2;
        if (i != max1_index && v[i] > max2) {
            max2 = v[i];
            ins++;
        }
    }

    ins++;
    return max1 * max2;
}

int main() {
    int n;
    ins++;
    cin >> n;
    ins++;

    vector<int> v(n);
    ins += n;

    for (int i = 0; i < n; ++i) {
        ins += 2;
        cin >> v[i];
        ins++;
    }

    long long result = max_pairwise_product_f(v);
    cout << result << '\n';
    cout << "Executed instructions: " << ins << '\n';

    return 0;
}

Writing max_pairwise_product_ef_ins.cpp


Calculate estimated time by instruction and number of instructions by second

In [132]:
!g++ max_pairwise_product_ef_ins.cpp -o max_pairwise_product_ef_ins
!./max_pairwise_product_ef_ins < input1.txt
!./max_pairwise_product_ef_ins < input2.txt
!./max_pairwise_product_ef_ins < input3.txt

90
Executed instructions: 115
90
Executed instructions: 89
40000000000
Executed instructions: 64


In [140]:
%%writefile max_pairwise_product_ef_ins_py.py
ins = 0
def max_pairwise_product_f(l):
    global ins
    s = 0
    ins += 1
    m = 0
    ins += 1
    ins += 1
    for i in l:
        ins += 2
        if i > m:
            m = i
            ins += 1
        if i > s:
            m = s
            ins += 1
            s = i
            ins += 1

    ins += 2
    return s * m

if __name__ == "__main__":
        n = int(input())
        ins += 1
        l = list(map(int, input().split()))
        ins += 5 * n
        ins += 1
        print(max_pairwise_product_f(l))
        print(f"Executed instructions: {ins}")

Overwriting max_pairwise_product_ef_ins_py.py


In [141]:
!python max_pairwise_product_ef_ins_py.py < input1.txt
!python max_pairwise_product_ef_ins_py.py < input2.txt

90
Executed instructions: 107
90
Executed instructions: 81


### Estimate and verify the maximun n that can be solved in 1 second

In [142]:
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  ofstream MyFile("input15.txt");

  int n = 50000000;

  MyFile << n << "\n";


  const int buffer_size = 1024 * 1024;
  char buffer[buffer_size];
  int buffer_pos = 0;

  for (int i = 1; i <= n; ++i) {
    int written = snprintf(buffer + buffer_pos, buffer_size - buffer_pos, "%d ", i);
    buffer_pos += written;

    if (buffer_pos >= buffer_size - 100) {
      MyFile.write(buffer, buffer_pos);
      buffer_pos = 0;
    }
  }

  if (buffer_pos > 0) {
    MyFile.write(buffer, buffer_pos);
  }

  MyFile.close();

  return 0;
}

Overwriting WriteTest.cpp


In [143]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [144]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef_cpp

In [145]:
!time ./max_pairwise_product_ef_cpp < input15.txt

2499999950000000

real	0m26.363s
user	0m25.696s
sys	0m0.315s


## Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


In [146]:
%%writefile max_pairwise_product_gef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

using namespace std;

int main() {
    int n;
    cin >> n;

    if (n < 2) {
        cout << 0 << "\n";
        return 0;
    }

    long long max1 = LLONG_MIN;
    long long max2 = LLONG_MIN;

    for (int i = 0; i < n; ++i) {
        long long x;
        cin >> x;

        if (x > max1) {
            max2 = max1;
            max1 = x;
        } else if (x > max2) {
            max2 = x;
        }
    }

    cout << max1 * max2 << "\n";
    return 0;
}

Writing max_pairwise_product_gef_cpp.cpp


Test with inputs 1 and 2




In [147]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!./max_pairwise_product_gef_cpp < input1.txt
!./max_pairwise_product_gef_cpp < input2.txt

90
90


Test and time it with input 3

In [148]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!time ./max_pairwise_product_gef_cpp < input3.txt

40000000000

real	0m0.003s
user	0m0.002s
sys	0m0.001s


20 stress tests

In [150]:
%%writefile stress_tests_gef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
#include <random>

using namespace std;

int main() {
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> distrib(2, 1001);

    for (int i = 0; i < 20; ++i) {
        int n = distrib(gen) % 77 + 2;

        long long max1 = LLONG_MIN;
        long long max2 = LLONG_MIN;

        for (int j = 0; j < n; ++j) {
            long long x = distrib(gen);

            if (x > max1) {
                max2 = max1;
                max1 = x;
            } else if (x > max2) {
                max2 = x;
            }
        }

        cout << "n: " << n << ", Product: " << max1 * max2 << endl;
    }

    return 0;
}

Writing stress_tests_gef_cpp.cpp


In [151]:
!g++ stress_tests_gef_cpp.cpp -o stress_tests_gef_cpp
!./stress_tests_gef_cpp

n: 29, Product: 964243
n: 51, Product: 977042
n: 76, Product: 929296
n: 40, Product: 900600
n: 26, Product: 922077
n: 3, Product: 25704
n: 52, Product: 997002
n: 73, Product: 969228
n: 28, Product: 947000
n: 28, Product: 986013
n: 23, Product: 465750
n: 2, Product: 69960
n: 60, Product: 949644
n: 38, Product: 967182
n: 55, Product: 904205
n: 8, Product: 364089
n: 66, Product: 966264
n: 5, Product: 811958
n: 34, Product: 841534
n: 51, Product: 965965


Count executed instructions in code and verify with inputs 1,2 and 3

In [152]:
%%writefile max_pairwise_product_gef_ins_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

using namespace std;

int main() {
    uint64_t ins = 0;
    int n;
    ins++;
    cin >> n;
    ins++;

    if (n < 2) {
        cout << 0 << "\n";
        cout << "Executed instructions: " << ins << '\n';
        return 0;
    }

    long long max1 = LLONG_MIN;
    ins++;
    long long max2 = LLONG_MIN;
    ins++;

    for (int i = 0; i < n; ++i) {
        ins += 2;
        long long x;
        ins++;
        cin >> x;
        ins++;

        if (x > max1) {
            max2 = max1;
            max1 = x;
            ins += 2;
        } else if (x > max2) {
            max2 = x;
            ins++;
        }
    }

    ins += 3;
    cout << max1 * max2 << "\n";
    cout << "Executed instructions: " << ins << '\n';

    return 0;
}

Writing max_pairwise_product_gef_ins_cpp.cpp


In [154]:
!g++ max_pairwise_product_gef_ins_cpp.cpp -o max_pairwise_product_gef_ins_cpp
!time ./max_pairwise_product_gef_ins_cpp < input1.txt
!time ./max_pairwise_product_gef_ins_cpp < input2.txt
!time ./max_pairwise_product_gef_ins_cpp < input3.txt


90
Executed instructions: 67

real	0m0.005s
user	0m0.002s
sys	0m0.001s
90
Executed instructions: 50

real	0m0.003s
user	0m0.002s
sys	0m0.001s
40000000000
Executed instructions: 39

real	0m0.004s
user	0m0.001s
sys	0m0.002s


Calculate estimated time by instruction and number of instructions by second

In [ ]:
instructions_per_second = 0.005 / 67
time_per_instruction = 67 / 0.005

print(f"Estimated time by instruction: {time_per_instruction} seconds")
print(f"Estimated number of instructions by second: {instructions_per_second}")

### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [155]:
%%writefile max_pairwise_product_gef_py.py
#Python3
def max_pairwise_product_gef(numbers):
    n = len(numbers)
    if n < 2:
        return 0

    max1 = float('-inf')
    max2 = float('-inf')

    for x in numbers:
        if x > max1:
            max2 = max1
            max1 = x
        elif x > max2:
            max2 = x

    return max1 * max2

if __name__ == "__main__":
    n = int(input())
    numbers = list(map(int, input().split()))
    print(max_pairwise_product_gef(numbers))

Writing max_pairwise_product_gef_py.py


Test with inputs 1 and 2




In [156]:
!python max_pairwise_product_gef_py.py < input1.txt
!python max_pairwise_product_gef_py.py < input2.txt

90
90


Test and time it with input 3

In [157]:
%%time
!python max_pairwise_product_gef_py.py < input3.txt

40000000000
CPU times: user 5.43 ms, sys: 831 µs, total: 6.26 ms
Wall time: 108 ms


20 stress tests

In [158]:
%%writefile stress_tests_gef_py.py
import random

def max_pairwise_product_gef(numbers):
    n = len(numbers)
    if n < 2:
        return 0

    max1 = float('-inf')
    max2 = float('-inf')

    for x in numbers:
        if x > max1:
            max2 = max1
            max1 = x
        elif x > max2:
            max2 = x

    return max1 * max2

if __name__ == "__main__":
    for _ in range(20):
        n = random.randint(2, 99)
        numbers = [random.randint(1, 150000) for _ in range(n)]
        product = max_pairwise_product_gef(numbers)
        print(f"n: {n}, Product: {product}")

Writing stress_tests_gef_py.py


In [159]:
!python stress_tests_gef_py.py

n: 5, Product: 10544991352
n: 45, Product: 22127567960
n: 24, Product: 21825034560
n: 83, Product: 21677785025
n: 59, Product: 22129048963
n: 44, Product: 21181303515
n: 27, Product: 21481024100
n: 2, Product: 1921729536
n: 3, Product: 4630858941
n: 31, Product: 21137770616
n: 11, Product: 21448123770
n: 92, Product: 22219237040
n: 80, Product: 21155619794
n: 63, Product: 19937329077
n: 66, Product: 19071011376
n: 90, Product: 21399115980
n: 53, Product: 22359030364
n: 49, Product: 21833718408
n: 78, Product: 21102947040
n: 55, Product: 21424937664


Count executed instructions in code and verify with inputs 1,2 and 3

In [160]:
%%writefile max_pairwise_product_gef_ins_py.py
def max_pairwise_product_gef_ins(numbers):
    ins = 0
    n = len(numbers)
    ins += 1

    if n < 2:
        return 0, ins

    max1 = float('-inf')
    ins += 1
    max2 = float('-inf')
    ins += 1

    for x in numbers:
        ins += 1
        if x > max1:
            max2 = max1
            max1 = x
            ins += 2
        elif x > max2:
            max2 = x
            ins += 1

    ins += 1
    return max1 * max2, ins

if __name__ == "__main__":
    ins = 0
    n = int(input())
    ins += 1
    numbers = list(map(int, input().split()))
    ins += 5 * n + 1
    product, ins_count = max_pairwise_product_gef_ins(numbers)
    ins += ins_count
    ins += 2
    print(product)
    print(f"Executed instructions: {ins}")

Writing max_pairwise_product_gef_ins_py.py


In [162]:
%%time
!python max_pairwise_product_gef_ins_py.py < input1.txt
!python max_pairwise_product_gef_ins_py.py < input2.txt
!python max_pairwise_product_gef_ins_py.py < input3.txt

90
Executed instructions: 88
90
Executed instructions: 71
40000000000
Executed instructions: 52
CPU times: user 15.7 ms, sys: 2.15 ms, total: 17.8 ms
Wall time: 318 ms


Calculate estimated time by instruction and number of instructions by second

In [163]:
number_of_ins_per_second = 80 / 0.0178
estimated_time = 80 / 4494.38
print(f"Estimated time by instruction: {estimated_time} seconds")
print(f"Estimated number of instructions by second: {number_of_ins_per_second}")

Estimated time by instruction: 0.017800008010003605 seconds
Estimated number of instructions by second: 4494.38202247191


### Estimate and verify the maximun n that can be solved in 1 second

In [164]:
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  ofstream MyFile("input15.txt");

  int n = 207715727;

  MyFile << n << "\n";
  const int buffer_size = 1024 * 1024;
  char buffer[buffer_size];
  int buffer_pos = 0;

  for (int i = 1; i <= n; ++i) {
    int written = snprintf(buffer + buffer_pos, buffer_size - buffer_pos, "%d ", i);
    buffer_pos += written;

    if (buffer_pos >= buffer_size - 100) {
      MyFile.write(buffer, buffer_pos);
      buffer_pos = 0;
    }
  }
  if (buffer_pos > 0) {
    MyFile.write(buffer, buffer_pos);
  }

  MyFile.close();

  return 0;
}

Overwriting WriteTest.cpp


In [165]:
!g++ WriteTest.cpp -o WriteTest

In [166]:
!./WriteTest

^C


In [167]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!time ./max_pairwise_product_gef_cpp < input15.txt

169022074720801

real	0m6.457s
user	0m6.418s
sys	0m0.032s


## Perform 100 randomized automated tests in C++ and Python with a porgram  compares the answers,for the same 100 random input instances, of the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n \lg{n}))$,  if the answers are not the same stops and displays "Error" listing  the input and the two answers; if the 100 tests are completed without stoping displays "Pass".

In [170]:
%%writefile test_max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>
#include <climits>

using namespace std;
long long max_pairwise_product_fast(const vector<int>& numbers) {
    int n = numbers.size();
    if (n < 2) {
        return 0;
    }

    long long max1 = LLONG_MIN;
    long long max2 = LLONG_MIN;

    for (long long x : numbers) {
        if (x > max1) {
            max2 = max1;
            max1 = x;
        } else if (x > max2) {
            max2 = x;
        }
    }

    return max1 * max2;
}
long long max_pairwise_product_slow(vector<int>& numbers) {
    int n = numbers.size();
    if (n < 2) {
        return 0;
    }
    sort(numbers.begin(), numbers.end());
    return (long long)numbers[n - 1] * numbers[n - 2];
}

int main() {
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> distrib(1, 100000);

    for (int i = 0; i < 100; ++i) {
        int n = distrib(gen) % 100 + 2; // Random input size between 2 and 101
        vector<int> numbers(n);
        for (int j = 0; j < n; ++j) {
            numbers[j] = distrib(gen); // Random input values
        }

        long long res1 = max_pairwise_product_fast(numbers);
        long long res2 = max_pairwise_product_slow(numbers);

        if (res1 != res2) {
            cout << "Error: Results don't match!" << endl;
            cout << "Input: ";
            for (int x : numbers) {
                cout << x << " ";
            }
            cout << endl;
            cout << "Fast solution: " << res1 << endl;
            cout << "Slow solution: " << res2 << endl;
            return 1; // Exit with error code
        }
    }

    cout << "Pass" << endl; // All tests passed
    return 0;
}

Overwriting test_max_pairwise_product.cpp


In [171]:
!g++ test_max_pairwise_product.cpp -o test_max_pairwise_product

In [172]:
!./test_max_pairwise_product

Pass


In [175]:
%%writefile stress_test_max_pairwise_product.py
import random

def max_pairwise_product_fast(numbers):
    n = len(numbers)
    if n < 2:
        return 0

    max1 = float('-inf')
    max2 = float('-inf')

    for x in numbers:
        if x > max1:
            max2 = max1
            max1 = x
        elif x > max2:
            max2 = x

    return max1 * max2

def max_pairwise_product_slow(numbers):
    n = len(numbers)
    if n < 2:
        return 0

    numbers.sort()
    return numbers[n - 1] * numbers[n - 2]

def stress_test(num_tests):
    for _ in range(num_tests):
        n = random.randint(2, 101)
        numbers = [random.randint(1, 100000) for _ in range(n)]

        res1 = max_pairwise_product_fast(numbers)
        res2 = max_pairwise_product_slow(numbers)

        if res1 != res2:
            print("Error: Results don't match!")
            print(f"Input: {numbers}")
            print(f"Fast solution: {res1}")
            print(f"Slow solution: {res2}")
            return False

    print("Pass")
    return True

if __name__ == "__main__":
    stress_test(100)

Writing stress_test_max_pairwise_product.py


In [176]:
! python stress_test_max_pairwise_product.py

Pass
